modules/metrics/src/main/java/com/spotify/apollo/metrics/MetricsCollectingEndpointRunnableFactoryDecorator.java

```diff
@@ -42,7 +42,20 @@
   @Override
   public EndpointRunnableFactory apply(EndpointRunnableFactory delegate) {
     return (request, requestContext, endpoint) -> {
-      final String endpointName = endpoint.info().getName();
+      final String endpointName;
+      // Normally request.request().method() and endpoint.info().getRequestMethod() are equal
+      // but since apollo lets GET endpoints also serve HEAD requests, they are different
+      // when the request is a HEAD request. The hack below makes sure we distinguish metrics
+      // for GET and HEAD requests.
+      // We could always take the else branch but I'm guessing that the comparison
+      // is cheaper than a needless concatenation.
+      if (request.request().method().equals(endpoint.info().getRequestMethod())) {
+        endpointName = endpoint.info().getName();
+      } else {
+        // but because GET endpoints serve HEAD requests info().getName() is
+        // wrong does not distinguish HEAD and GET requests
+        endpointName = request.request().method() + ":" + endpoint.info().getUri();
```

Comment:
Looks like this could work. But it would be more elegant if proper endpoint infos were set up for the implicit HEAD endpoints. This is a bit of a hack, isn't it?

Reply:
Definitely a hack. I'll see if I can come up with a better option.